hal/src/thumbv6m/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1353 @@
+    #[inline]
+    pub unsafe fn read_data(&mut self) -> u16 {
+        self.sercom().usart().data.read().data().bits()
+    }
+
+    /// Read the status register and convert into a [`Result`]
+    /// containing the corresponding [`Flags`] or [`Error`]
+    #[inline]
+    fn read_flags_errors(&self) -> Result<Flags, Error> {
+        self.read_status().try_into()?;
+        Ok(self.read_flags())
+    }
+
+    /// Flush the RX buffer and clear errors
+    #[inline]
+    pub fn flush(&mut self) {
+        let usart = unsafe { self.sercom().usart() };
+
+        // TODO
+        // The datasheet states that disabling the receiver (RXEN) clears
+        // the RX buffer, and clears the BUFOVF, PERR and FERR bits.
+        // However, in practice, it seems like BUFOVF errors still pop
+        // up after a disable/enable cycle of the receiver, then immediately begin
+        // reading bytes from the DATA register.
+        //
+        // Is this a hardware bug???
+        /*
+        usart.ctrlb.modify(|_, w| w.rxen().clear_bit());
+        while usart.syncbusy.read().ctrlb().bit() || usart.ctrlb.read().rxen().bit_is_set() {}
+
+        usart.ctrlb.modify(|_, w| w.rxen().set_bit());
+        while usart.syncbusy.read().ctrlb().bit() || usart.ctrlb.read().rxen().bit_is_clear() {}
+        */
+
+        // Workaround is to read a few bytes to clear the RX buffer (3 bytes seems to do
+        // the trick), then manually clear the BUFOVF bit Clear rx buffer
+        for _ in 0..=2 {
+            let _data = usart.data.read();
+        }
+
+        // Clear all errors
+        self.clear_status(
+            Status::BUFOVF | Status::FERR | Status::PERR | Status::ISF | Status::COLL,
+        );
+    }
+}
+
+impl<C, D> Uart<C, D>
+where
+    C: ValidConfig,
+    D: Transmit,
+{
+    /// Write to the DATA register
+    ///
+    /// # Safety
+    ///
+    /// Writing to the data register directly is `unsafe`, because it will clear
+    /// the DRE flag, which could break assumptions made elsewhere in this
+    /// module.
+    #[inline]
+    pub unsafe fn write_data(&mut self, data: u16) {
+        self.sercom().usart().data.write(|w| w.data().bits(data))
+    }
+}
+
+//=============================================================================
+// Embedded HAL traits
+//=============================================================================
+
+impl<C, D> Read<C::Word> for Uart<C, D>
+where
+    C: ValidConfig,
+    C::Word: PrimInt,
+    D: Receive,
+    u16: AsPrimitive<C::Word>,
+{
+    type Error = Error;
+
+    /// Wait for an `RXC` flag, then read the word
+    #[inline]
+    fn read(&mut self) -> nb::Result<C::Word, Error> {
+        let flags = self.read_flags_errors()?;
+        if flags.contains(Flags::RXC) {
+            unsafe { Ok(self.read_data().as_()) }
+        } else {
+            Err(WouldBlock)
+        }
+    }
+}
+
+impl<C, D> Write<C::Word> for Uart<C, D>
+where
+    C: ValidConfig,
+    C::Word: PrimInt + AsPrimitive<u16>,
```

Comment:
You can also add `AsPrimitive<u16>` to `CharSize::Word`. You can add `u16: AsPrimitive<Word>` as well, but it won't be inferred anywhere, so you'll still need to propagate the bound.